application/F3DStarter.cxx

```diff
@@ -330,47 +332,168 @@ void F3DStarter::LoadFile(f3d::loader::LoadFileEnum load)
+          this->Internals->LoadedFile = true;
+        }
+        else if (loader.hasGeometryReader(filePath.string()))
+        {
+          // In GroupGeometries, just load all the files from the list
+          if (fileAppOptions.GroupGeometries)
+          {
+            int nGeom = 0;
+            for (size_t i = 0; i < size; i++)
+            {
+              auto geomPath = this->Internals->FilesList[i];
+              if (loader.hasGeometryReader(geomPath.string()))
+              {
+                // Reset for the first file, then add geometries without resetting
+                loader.loadGeometry(this->Internals->FilesList[i].string(), i == 0 ? true : false);
+                nGeom++;
+              }
+              else
+              {
+                f3d::log::warn(geomPath, " is not a geometry of a supported file format\n");
+              }
+            }
+            if (nGeom > 1)
+            {
+              filenameInfo = std::to_string(nGeom) + " geometries loaded";
+            }
+          }
+          else
+          {
+            // Standard loadGeometry code
+            loader.loadGeometry(filePath.string(), true);
+          }
+          this->Internals->LoadedFile = true;
+        }
+        else
+        {
+          f3d::log::warn(filePath, " is not a file of a supported file format\n");
+          filenameInfo += " [UNSUPPORTED]";
+        }
+      }
+      catch (const f3d::loader::load_failure_exception& ex)
+      {
+        // XXX Not reachable until vtkImporter is improved to support returning a failure
+        f3d::log::error("Could not load file: ", ex.what());
+      }
+
+      if (!this->Internals->AppOptions.NoRender)
+      {
+        // Setup the camera according to options
+        this->Internals->SetupCamera(fileAppOptions);
+
+        this->Internals->Engine->getWindow().setWindowName(
+          filePath.filename().string() + " - " + F3D::AppName);
+      }
     }
   }
+  this->Internals->Engine->getOptions().set("ui.filename-info", filenameInfo);
 }
 
 //----------------------------------------------------------------------------
 void F3DStarter::Render()
 {
   this->Internals->Engine->getWindow().render();
 }
+
+//----------------------------------------------------------------------------
+void F3DStarter::AddFile(const fs::path& path)
+{
+  auto tmpPath = fs::absolute(path);
+  if (!fs::exists(tmpPath))
+  {
+    f3d::log::error("File ", tmpPath, " does not exist");
+    return;
+  }
+
+  if (fs::is_directory(tmpPath))
+  {
+    std::set<fs::path> sortedPaths;
+    for (const auto& entry : fs::directory_iterator(tmpPath))
+    {
+      sortedPaths.insert(entry.path());
+    }
+    for (auto& entryPath : sortedPaths)
+    {
+      // Recursively add all files
+      this->AddFile(entryPath);
+    }
+  }
+  else
+  {
+    auto it =
+      std::find(this->Internals->FilesList.begin(), this->Internals->FilesList.end(), tmpPath);
+
+    if (it == this->Internals->FilesList.end())
```

Comment:
should we warn the user the file is already added?

Reply:
Yes, or at least we should switch to it anyway, even if already added. I will make the change.

Reply:
actually that makes it way too complex with the need of index tracking. I will just emit a warning.